src/Perl6/Metamodel/Dispatchers.nqp

```diff
@@ -1,32 +1,64 @@
 class Perl6::Metamodel::BaseDispatcher {
     has @!candidates;
     has $!idx;
+    has $!next_dispatcher; # The dispatcher we must pass control to when own queue exhausts
 
     method candidates() { @!candidates }
 
-    method exhausted() { $!idx >= +@!candidates }
+    method exhausted() { $!idx >= +@!candidates && (!nqp::isconcrete($!next_dispatcher) || $!next_dispatcher.exhausted()) }
 
     method last()      { @!candidates := [] }
 
+    method set_next_dispatcher ($next_dispatcher) { $!next_dispatcher := $next_dispatcher }
```

Comment:
Completely unrelated to what this PR actually does, but you have a couple methods with a space between the method name and the signature.

Reply:
My Perl6 habbit. Will fix now. Thanks!